File: subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/cegar/astar/argstore/AstarArgStore.java
package hu.bme.mit.theta.analysis.algorithm.cegar.astar.argstore;

import hu.bme.mit.theta.analysis.Action;
import hu.bme.mit.theta.analysis.Prec;
import hu.bme.mit.theta.analysis.State;
import hu.bme.mit.theta.analysis.algorithm.cegar.astar.AstarArg;

public interface AstarArgStore<S extends State, A extends Action, P extends Prec> {
	void add(AstarArg<S, A, P> astarArg);
	AstarArg<S, A, P> get(int index);
	AstarArg<S, A, P> getLast();
	int size();
	int getIndex(AstarArg<S, A, P> astarArg);
}

File: subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/cegar/astar/argstore/AstarArgStore.kt
package hu.bme.mit.theta.analysis.algorithm.cegar.astar.argstore

import hu.bme.mit.theta.analysis.Action
import hu.bme.mit.theta.analysis.Prec
import hu.bme.mit.theta.analysis.State
import hu.bme.mit.theta.analysis.algorithm.cegar.astar.AstarArg

interface AstarArgStore<S: State, A: Action, P: Prec> {
	val size: Int

	fun add(astarArg: AstarArg<S, A, P>): Boolean

	operator fun get(index: Int): AstarArg<S, A, P>

	val last: AstarArg<S, A, P>
		get() = get(size - 1)

	fun indexOf(astarArg: AstarArg<S, A, P>): Int

	fun setLast(astarArg: AstarArg<S, A, P>)

	operator fun plusAssign(astarArg: AstarArg<S, A, P>) {
		this.add(astarArg)
	}
}
